File: apps/desktop/core/galaxyssi-link/backend/mqtt_chunk_feedback.py
"""Bounded latest-state feedback, flushed by the existing pool maintenance tick."""
from dataclasses import dataclass
import threading

from mqtt_broker_catalog import CATALOG


@dataclass
class Pending:
    due: float
    revision: int
    send: object


class ChunkFeedback:
    def __init__(self):
        self.pending = {}
        self.lock = threading.Lock()

    def offer(self, scope, transfer, request, revision, send, now, urgent=False):
        key = scope, transfer, request
        with self.lock:
            previous = self.pending.get(key)
            if previous and revision < previous.revision:
                return None
            if urgent:
                self.pending.pop(key, None)
                return send
            if previous is None and (len(self.pending) >= CATALOG["limits"]["max_chunk_feedback"]
                    or sum(item[0] == scope for item in self.pending) >= CATALOG["limits"]["per_peer_chunk_feedback"]):
                return None
            due = previous.due if previous else now + CATALOG["timing"]["chunk_feedback_window_ms"] / 1000
            self.pending[key] = Pending(due, revision, send)
        return None

    def drain(self, now, limit=16):
        with self.lock:
            keys = [key for key, value in self.pending.items() if value.due <= now][:limit]
            return [self.pending.pop(key).send for key in keys]

    def forget(self, scope):
        with self.lock:
            self.pending = {key: value for key, value in self.pending.items() if key[0] != scope}

File: apps/android/app/src/main/java/com/galaxyssi/chat/MqttChunkFeedback.kt
package com.galaxyssi.chat

/** Coalesced state only. Existing MQTT maintenance owns flushing; no new worker or timer. */
internal class MqttChunkFeedback {
    private data class Key(val scope: String, val transfer: String, val request: String)
    private data class Pending(val due: Long, val revision: Long, val send: () -> Unit)
    private val pending = linkedMapOf<Key, Pending>()

    @Synchronized fun offer(scope: String, transfer: String, request: String, revision: Long, send: () -> Unit,
                            now: Long, urgent: Boolean = false): (() -> Unit)? {
        val key = Key(scope, transfer, request)
        val previous = pending[key]
        if (previous != null && revision < previous.revision) return null
        if (urgent) { pending.remove(key); return send }
        if (previous == null && (pending.size >= MqttBrokerCatalog.MAX_CHUNK_FEEDBACK ||
                pending.keys.count { it.scope == scope } >= MqttBrokerCatalog.PEER_CHUNK_FEEDBACK)) return null
        pending[key] = Pending(previous?.due ?: (now + MqttBrokerCatalog.CHUNK_FEEDBACK_WINDOW_MS), revision, send)
        return null
    }
    @Synchronized fun drain(now: Long, limit: Int = 16): List<() -> Unit> = pending.filterValues { it.due <= now }
        .keys.take(limit).map { pending.remove(it)!!.send }
    @Synchronized fun forget(scope: String) { pending.keys.removeAll { it.scope == scope } }
}
